### PR TITLE
Fix tag-finder for self-closing tags xml tags (<foo/>)

### PR DESCRIPTION
**Pull request by MoritzKn**
### Requirements

* [x] Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* [x] All new code requires tests to ensure against regressions

### Description of the Change
Adjust the tag regex used to find not closed tags to ignore self-closing tags.

### Alternate Designs
None

### Benefits
Will fix a bug (described in #203)

### Possible Drawbacks
None

### Applicable Issues
#203

### CODE REVIEW

**Comment by 50Wliu** [comment]
@MoritzKn there is already an open PR for #203: #257.  Is there anything you do in this PR that could be considered "superior" to the approach taken in #257?

**Comment by MoritzKn** [comment]
@50Wliu 
**TL;DR:**
both PRs should be accepted. 

I noticed that both PRs fix different aspects of the problem with self closing tags:
- This PR fixes the problems with the functions `tagsNotClosedInFragment` and `tagsNotClosedInFragment` that are used for the `bracket-matcher:close-tag` command.
- #257 Fixes the tag highlighting of corresponding tags.

Both PRs can easily be merged; they have no conflicts.

### Examples
In this file
```xml
<foo>
    <bar/>
[cursor here]
```
Calling `bracket-matcher:close-tag` will insert
-  `</bar>` with out any PR applied
-  `</bar>` with #257 applied
-  `</foo>` with this PR applied
-  `</foo>` with this PR and #257 applied

On the other hand in this example:
```xml
<bar>
    <bar/>
</bar[cursor here]> 
```
-  `</bar>` and `<bar/>` will be highlighted with out any PR applied
-  `</bar>` and `<bar/>` will be highlighted with #257 applied
-  `</bar>` and `<bar>` will be highlighted with this PR applied
-  `</bar>` and `<bar>` will be highlighted with  #257 and this PR applied

**Comment by ruv** [comment]
Don't quite understand what review we are waiting for? The changes are pretty short and simple.

Actually, attributes in a self-closing tag can contain unescaped '>' symbol, so this fix doesn't cover all the cases, but it is better than nothing.

Testcases for uncovered case: `<bar test="a>b" />`  and `<bar test="a/>b" >`

**Comment by ruv** [comment]
I suggest the following improvement to fix the mentioned issue with possible unescaped '>' symbol in attributes:
`(?:[^'"\/>]*(?:"[^"]*"|'[^']*'))*\s*>` in place of `(?![^>]*\/>)`
I.e. we just match all attributes up to '>' at the end of a tag.

In general, it is better to compose such complex regex using string interpolations (templates). Something like:
`re = new RegExp("<(#{name})(?:#{attribute})*\s*>|</(#{name})\s*>")`

**Comment by MoritzKn** [comment]
@rur Year I was thinking about something like this too but I think your suggestion introduces a problem with empty attributes like `<video autoplay>`. I now ended up with `(?!(?:"[^"]*"|'[^']*'|[^>])*\/>)`, it's also a little bit simpler. Let me know if you see a problem with this variant.

**Comment by ruv** [comment]
Yes, I have edited the message and updated the regex to also match html attributes (i.e. just a name without value) — just add one '?'.

In any case, yes — your variant is simpler. But why do you use negated lookahead? It is sub-optimal — this fragment will be analyzed twice instead of skipping. I suggest: `(?:"[^"]*"|'[^']*'|[^>\/])*>`

**Comment by MoritzKn** [comment]
@ruv 
> But why do you use negated lookahead? It is sub-optimal — this fragment will be analyzed twice instead of skipping.

Valid point. I will follow your suggestion, but for some reason I had to add `"'` to the `[^>\/]` match group to get it working so I now ended up with this regex: 
```regexp
(?:"[^"]*"|'[^']*'|[^>\/"'])*>
```

**Comment by ruv** [comment]
> for some reason I had to add `"'` to the `[^>\/]` match group 

Could you please show an example, why do we need this change?

**Comment by MoritzKn** [comment]
@ruv sure: https://regex101.com/r/Kl3lO7/1

**Comment by ruv** [comment]
Thanks, I see, — I have forgotten about greedy backtracking ;)

**Comment by ruv** [comment]
For optimization we can just change order:
```regex
(?:[^>\/"']|"[^"]*"|'[^']*')*>
```

**Comment by MoritzKn** [comment]
I updated the PR. But as described in the commit message, the decision to not use a negative lookahead broke a spec. But because [the old behavior was already criticized](https://github.com/atom/bracket-matcher/pull/229/files#diff-96af432f5b62f03fb0fb58e2f4f01310) I just removed the spec. If the old behavior is still desired I can change the regex and restore the spec.
I also update the doc comment of the tagsNotClosedInFragment function, this has nothing to do with this PR, I just noted that it was outdated.

**Comment by ungb** [comment]
I've tested this pr with html file and saw that this fixes the closing tag issue.  

working

**Comment by as-cii** [comment]
This seems reasonable, thanks for your pull request @MoritzKn! 🙇
